util, config: error on startup if `conf` or `reindex` are set in config file

josibake opened the pull request:
In help from `bitcoind -h` it specifes that `conf` can only be used from the commandline. However, if `conf` is set in a `bitcoin.conf` file, there is no error and from reading the logs it seems as if the `conf=<other file>` is being used, despite it being ignored. To recreate, you can setup a `bitcoin.conf` file in the default directory, add `conf=<some other file>.conf` and in the separate config file set whichever config value you want and verify that it is being ignored. alternatively, if you set `includeconf=<some other file>.conf` , your config in `<some other file>` will be picked up.

This PR fixes this by having the node error when reading the config file if `conf=` is set.

Additionally, it was mentioned in a recent [PR review club](https://bitcoincore.reviews/24858) that if `reindex=1` is set in the config file, the node will reindex on every startup, which is undesirable:
 ```irc
17:14 <larryruane> michaelfolkson: Reindex is requested by the user (node operator) as a configuration option (command line or in the config file, tho you probably would never put it in the file, or else it would reindex on every startup!)
```



This PR also has a commit to warn if `reindex=1` is set in the config file.

josibake committed:
util: disallow setting conf in bitcoin.conf

Help from `bitcoind -h` states that conf can only be used from the commandline.
However, if conf is set in a bitcoin.conf file, it is ignored but there is no error.

Show an error to user if conf is set in a .conf file and prompt them to use
`includeconf` if they wish to specify additional config files.

Adds `IsConfSupported` function to allow for easily adding conf options
to disallow or throw warnings for.

aureleoules commented on src/init.cpp:
```diff
@@ -430,7 +430,7 @@ void SetupServerArgs(ArgsManager& argsman)
             "Warning: Reverting this setting requires re-downloading the entire blockchain. "
             "(default: 0 = disable pruning blocks, 1 = allow manual pruning via RPC, >=%u = automatically prune block files to stay under the specified target size in MiB)", MIN_DISK_SPACE_FOR_BLOCK_FILES / 1024 / 1024), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
-    argsman.AddArg("-reindex", "Rebuild chain state and block index from the blk*.dat files on disk. This will also rebuild active optional indexes.", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-reindex", "Rebuild chain state and block index from the blk*.dat files on disk. This will also rebuild active optional indexes (only useable from command line, not configuration file)", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
```
You removed the period but I realize the punctuation is not consistent across the commands

josibake replied:
yeah, it's mixed in the file, but im used to (and prefer) not having punctuation in stdout help text

aureleoules replied:
you also mixed it here 😄
https://github.com/bitcoin/bitcoin/blob/79e5606e750b28fa2123f96ff07757c931ae5f91/src/util/system.cpp#L948-L955

josibake replied:
oops! good catch, ill make them consistent

aureleoules reviewed (commented):
ACK 79e5606e750b28fa2123f96ff07757c931ae5f91.

I think it is reasonable to throw an error if `conf` is set in the config file as it doesn't make sense.
At first I thought showing a warning if -reindex=1 is set in the config would be better than an error but I don't see any use-case for reindexing on every startup so I believe error-ing is better.

josibake commented:
force pushed to format error messages consistently; `git range-diff master 79e5606 c7ebf00`

aureleoules reviewed (commented):
re-ACK c7ebf00cdf1e9f275a847e07c3d7c6c940ab9cb2

furszy commented on src/util/system.cpp:
```diff
@@ -945,6 +945,14 @@ bool ArgsManager::ReadConfigStream(std::istream& stream, const std::string& file
     for (const std::pair<std::string, std::string>& option : options) {
         KeyInfo key = InterpretKey(option.first);
         std::optional<unsigned int> flags = GetArgFlags('-' + key.name);
+        if (key.name == "conf") {
+            error = "conf cannot be set in the configuration file; if you want to include additional configuration files, use includeconf";
+            return false;
+        }
+        if (key.name == "reindex") {
+            error = "reindex cannot be set in the configuratoin file; use -reindex=1 from the commandline";
+            return false;
+        }
```
quick thing:
 would be better to decouple this into a different function `IsConfKeySupported` (or something similar) to keep `ArgsManager::ReadConfigStream` as general as possible and not pollute it with the inclusion of every single not supported option (might be more in the future).

josibake replied:
great suggestion; added!

ryanofsky commented:
I think it would be good to log a warning if reindex is specified in the configuration file, but still allow it. It seems to me there may be a lot of cases where it is awkward or difficult to change command line arguments and easier to temporarily add or uncomment `reindex=1` from the config file. Examples might be when bitcoind is running as a systemd service (especially on a system with an immutable or autogenerated /etc) or where bitcoin-qt is started from an icon shortcut with hardcoded arguments.

Also I doubt reindex=1 is the only configuration option that most people won't want in their config file, so making it a hard error and treating it so differently than other configurations options seems a bit of a scattershot solution.

josibake commented:
> Examples might be when bitcoind is running as a systemd service (especially on a system with an immutable or autogenerated /etc)

That's a great point, I'll change it to a warning. Initially, I was worried a warning would go unnoticed by most users, although thinking about this a bit more, if my node were to take forever on every restart, most users would likely check the logs first and see the warning.

josibake committed:
util: warn if reindex is used in conf

using reindex in a conf file can lead to the node reindexing on every restart.
we still allow it but throw a warning.

josibake commented:
updated to only warn for `reindex` per @ryanofsky and added `IsConfSupported` function per @furszy

ryanofsky commented on src/util/system.cpp:
```diff
@@ -940,6 +940,12 @@ bool IsConfSupported(KeyInfo key, std::string& error) {
+    if (key.name == "reindex") {
+        // reindex can be set in a config file but it is strongly discouraged as this will cause the node to reindex on
+        // every restart. Allow the config but throw a warning
+        LogPrintf("Warning: reindex is set which will cause this node to reindex on every restart. Consider using -reindex=1 from the command-line\n");
```
In commit "util: warn if reindex is used in conf" (5cf1b465f4f9352415604c3eb31fc8cd01e4bd46)

This wording is not good because it doesn't clearly state the cause of the problem. It makes it sound like just setting reindex in general causes reindexing every restart. Also it makes it sound like adding -reindex=1 to the command line would fix the problem, when it wouldn't and the actual fix is to *remove* reindex=1 from the config file.

Would suggest: "Warning: reindex=1 is set in the configuration file, which will significantly slow down startup. Consider removing or commenting out this option for better performance, unless there is currently a condition which makes rebuilding the indexes necessary."

josibake replied:
nice, i like the suggested wording much better.

ryanofsky reviewed (approved):
Code review ACK 5cf1b465f4f9352415604c3eb31fc8cd01e4bd46. Code changes look good but I would consider improving the warning text, see below. Also it would be good to add either unit tests [[1]](https://github.com/bitcoin/bitcoin/blob/5871b5b5ab57a0caf9b7514eb162c491c83281d5/src/test/util_tests.cpp#L295) [[2]](https://github.com/bitcoin/bitcoin/blob/5871b5b5ab57a0caf9b7514eb162c491c83281d5/src/test/getarg_tests.cpp#L436) or functional tests [[3]](https://github.com/bitcoin/bitcoin/blob/5871b5b5ab57a0caf9b7514eb162c491c83281d5/test/functional/feature_config_args.py#L49) for this

EDIT: fixed second link

josibake committed:
test: update feature_config_args.py

add two new test cases for conf and reindex

josibake commented:
updated the wording per @ryanofsky 's suggestion and updated `test/functional/feature_config_args.py` to test for `conf` and `reindex`; `git range-diff master 5cf1b46 8f13984`

ryanofsky reviewed (approved):
Code review ACK 019e02cb26d87db9781353df8de70ed692069517. Thanks for update and new tests!

theStack commented:
Concept ACK

luke-jr reviewed (approved):
utACK

Note there is a downside to a warning for reindex: by the time the user sees it, it is too late to cancel and keep the old index.

1440000bytes reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/25727/commits/5f94b0210a8d4cf16f73f300c12d370af71bf9d8

Agreeing with only the changes in this commit

luke-jr commented on src/util/system.cpp:
```diff
@@ -935,6 +935,20 @@ static bool GetConfigOptions(std::istream& stream, const std::string& filepath,
     return true;
 }
 
+bool IsConfSupported(KeyInfo key, std::string& error) {
```
Maybe pass `key` by reference

tusher456 replied:
32sMEMSwnLLcuxzHptTDksk9auSPoccWiJ
My btc wallet send me some btc

1440000bytes replied:
@tusher456 please create an address in walllet on testnet. I will send you some BTC. Else you are asking at wrong place. This is not giveaway website.

josibake commented:
updated `IsConfSupported` to be pass by reference, per @luke-jr 's feedback; `git range-diff master 019e02c deba6fe`

hebasto commented on src/util/system.cpp:
```diff
@@ -935,6 +935,20 @@ static bool GetConfigOptions(std::istream& stream, const std::string& filepath,
     return true;
 }
 
+bool IsConfSupported(KeyInfo& key, std::string& error) {
```
As this function is used in the only compilation unit it could be declared within an anonymous `namespace` or as `static` to keep global namespace hygiene (the former option is my personal preference :tiger2: ).

See: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#sf22-use-an-unnamed-anonymous-namespace-for-all-internalnon-exported-entities

hebasto commented on src/util/system.cpp:
```diff
@@ -935,6 +935,20 @@ static bool GetConfigOptions(std::istream& stream, const std::string& filepath,
+    if (key.name == "reindex") {
+        // reindex can be set in a config file but it is strongly discouraged as this will cause the node to reindex on
+        // every restart. Allow the config but throw a warning
+        LogPrintf("Warning: reindex=1 is set in the configuration file, which will significantly slow down startup. Consider removing or commenting out this option for better performance, unless there is currently a condition which makes rebuilding the indexes necessary\n");
```
In a follow up, it would be nice to show this warning in the GUI as well.

hebasto reviewed (approved):
ACK deba6fe3158cd0b2283e0901a072e434ba5b594e, tested on Ubuntu 22.04.

aureleoules commented on src/util/system.cpp:
```diff
@@ -935,6 +935,14 @@ static bool GetConfigOptions(std::istream& stream, const std::string& filepath,
     return true;
 }
 
+bool IsConfSupported(KeyInfo& key, std::string& error) {
```
5e744f423838fe7d45453541271bc1a07cd62eac nit
```suggestion
bool IsConfSupported(const KeyInfo& key, std::string& error) {
```

aureleoules reviewed (approved):
tACK deba6fe3158cd0b2283e0901a072e434ba5b594e

ryanofsky reviewed (approved):
Code review ACK deba6fe3158cd0b2283e0901a072e434ba5b594e.

This looks good to merge, but I do agree with other comments the function should be static and KeyInfo reference should be const. Since IsConfSupported is a nice example of a function that returns errors and warnings, I think I will convert it to use `util::Result` later #25665, and I could make these cleanups then.

ryanofsky commented:
> Note there is a downside to a warning for reindex: by the time the user sees it, it is too late to cancel and keep the old index.

That's a fair point, but I think the point of the warning is to give users a clue of what is causing their node to take a long time to start if they accidentally misconfigured it, or turned on the reindex flag one time and forgot about it. If we make it an error we get rid of a convenient way of enabling reindexing, so the warning is compromise a way to help someone stuck with a node that starts slowly due to a misconfiguration but otherwise works, without inconveniencing a user who is setting reindex intentionally.

fanquake commented:
> and I could make these cleanups then.

@ryanofsky Sounds good.
